server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -242,59 +262,93 @@ public Object persist(final Committer committer) throws InterruptedException
   }
 
   /**
-   * Publish all data indexed through this driver so far, and waits for it to be handed off. Blocks until complete.
-   * Retries forever on transient failures, but may exit early on permanent failures.
-   *
-   * Should be called after all data has been added and persisted through {@link #add(InputRow, String, Supplier)} and
-   * {@link #persist(Committer)}.
+   * Register the segments in the given {@link SegmentsAndMetadata} to be handed off and execute a background task which
+   * waits until the hand off completes.
    *
-   * @param publisher publisher to use for this set of segments
-   * @param committer committer representing all data that has been added so far
-   *
-   * @return segments and metadata published if successful, or null if segments could not be handed off due to
-   * transaction failure with commit metadata.
+   * @param segmentsAndMetadata
+   * @return
+   * @throws InterruptedException
    */
-  public SegmentsAndMetadata finish(
-      final TransactionalSegmentPublisher publisher,
-      final Committer committer
-  ) throws InterruptedException
+  public ListenableFuture<SegmentsAndMetadata> registerHandoff(SegmentsAndMetadata segmentsAndMetadata)
```

Comment:
Yes as per the current design `get()` on Future returned by `registerHandoff()` would be same as `waitForHandOff`. However, in future when KafkaIndexTask has the capability to do incremental handOffs then having a method like `waitForHandOff` which just checks for whether all the segments handled by the appenderator irrespective of sequenceNames are handed Off or not will make things simpler. The IndexTask will not call this method whereas KafkaIndexTask can block on this method after all the publishes are done and before the task finishes, otherwise all the Futures returned by `registerHandoff()` need to be checked for completion in some thread while incorporating logic for `handoffConditionTimeout`. 

The current design also works but I feel having two methods is simpler, I guess its your choice whichever feels good to you.